SSA/Experimental/IntrinsicAsymptotics.lean

```diff
@@ -1358,3 +1358,142 @@ example : rewritePeepholeAt p4 5 ex3 = (
   .ret ⟨0, by simp⟩) := rfl
 
 end Examples
+
+namespace RegionExamples
+
+/-- A very simple type universe. -/
+inductive ExTy
+  | nat
+  deriving DecidableEq, Repr
+
+@[reducible]
+instance : Goedel ExTy where
+  toType
+    | .nat => Nat
+
+inductive ExOp :  Type
+  | add : ExOp
+  | runK : ℕ → ExOp
+  deriving DecidableEq
+
+instance : OpSignature ExOp ExTy where
+  outTy
+    | .add => .nat
+    | .runK _ => .nat
+  sig
+    | .add => [.nat, .nat]
+    | .runK _ => [.nat]
+  regSig
+   | .runK _ => [([.nat], .nat)]
+   | _ => []
+
+/-- Compose `f` with itself `n` times.
+  @chris: What's the mathlib correct version of this?
+ -/
+def loopN (f : α → α) (n : ℕ) (x: α) : α :=
+  match n with
+  | 0 => x
+  | n' + 1 => loopN f n' (f x)
+
+@[reducible]
+instance : OpDenote ExOp ExTy where
+  denote
+    | .add, .cons (a : Nat) (.cons b .nil), _ => a + b
+    | .runK (k : Nat), (.cons (v : Nat) .nil), (.cons rgn _nil) =>
+      loopN (n := k) (fun val => rgn (fun _ty _var => val)) v
+
+@[simp]
+def add {Γ : Ctxt _} (e₁ e₂ : Var Γ .nat) : IExpr ExOp Γ .nat :=
+  IExpr.mk
+    (op := .add)
+    (ty_eq := rfl)
+    (args := .cons e₁ <| .cons e₂ .nil)
+    (regArgs := .nil)
+
+@[simp]
+def rgn {Γ : Ctxt _} (k : Nat) (input : Var Γ .nat) (body : ICom ExOp [ExTy.nat] ExTy.nat) : IExpr ExOp Γ .nat :=
+  IExpr.mk
+    (op := .runK k)
+    (ty_eq := rfl)
+    (args := .cons input .nil)
+    (regArgs := HVector.cons body HVector.nil)
+
+macro "simp_peephole": tactic =>
+  `(tactic|
+      (
+      funext ll
+      simp only [ICom.denote, IExpr.denote, Var.zero_eq_last, Var.succ_eq_toSnoc,
+        Ctxt.snoc, Ctxt.Valuation.snoc_last, Ctxt.Valuation.snoc_toSnoc, add,
+        rgn, loopN,HVector.map, OpDenote.denote, IExpr.op_mk, IExpr.args_mk]
+      generalize ll { val := 0, property := _ } = a;
+      generalize ll { val := 1, property := _ } = b;
+      generalize ll { val := 2, property := _ } = c;
+      generalize ll { val := 3, property := _ } = d;
+      generalize ll { val := 4, property := _ } = e;
+      generalize ll { val := 5, property := _ } = f;
+      simp [Goedel.toType] at a b c d e f;
+      try clear f;
+      try clear e;
+      try clear d;
+      try clear c;
+      try clear b;
+      try clear a;
+      try revert f;
+      try revert e;
+      try revert d;
+      try revert c;
+      try revert b;
+      try revert a;
+      clear ll;
+      )
+   )
```

Comment:
I would prefer to see a single `simp_peephole` tactic that takes the context-specific simp-lemmas as argument.

So, things like `ICom.denote`, `IExpr.denote` can be part of the base `simp_peephole`, and then we can define the specialization to each context as:
```
macro "simp_reg_peephole" : tactic => `(tactic| simp_peephole [rgn, loopN, ...] )
```